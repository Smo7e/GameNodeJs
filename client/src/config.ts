// UdSU
// const DOMAIN = 'http://studfront';
// const PORT = 81;

// dev
const DOMAIN = "http://localhost";
const PORT = 3001;

// prod
//...

export const HOST = PORT ? `${DOMAIN}:${PORT}` : `${DOMAIN}`;

export const MEDIATOR = {
    EVENTS: {
        SERVER_ERROR: "SERVER_ERROR",
        GET_MESSAGES: "GET_MESSAGES",
        GET_SCENE: "GET_SCENE",
        GET_USER: "GET_USER",
        GET_GAMERS: "GET_GAMERS",
        GET_FRIENDS: "GET_FRIENDS",
        GET_INVITES: "GET_INVITES",
        LOGIN: "LOGIN",
        SIGNUP: "SIGNUP",
        LOGOUT: "LOGOUT",
        GET_MOBS: "GET_MOBS",
        GET_ERROR: "GET_ERROR",
        PLAY_MUSIC: "PLAY_MUSIC", 
        STOP_MUSIC: "STOP_MUSIC", 
        SET_MUSIC_VOLUME: "SET_MUSIC_VOLUME", 
        UPDATE_ARR_BULLET_TRAJECTORY: "UPDATE_ARR_BULLET_TRAJECTORY",
    },
    TRIGGERS: {},
};
