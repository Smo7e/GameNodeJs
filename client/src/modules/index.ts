import { TError } from "./Server/types";
import Server from "./Server/Server";
import Mediator from "./Mediator/Mediator";
import Store from "./Store/Store";

export type { TError };
export { Server, Mediator, Store };
