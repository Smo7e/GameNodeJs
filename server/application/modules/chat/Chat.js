class Chat {
    constructor(answer, db, io, mediator) {
        this.answer = answer;
        this.db = db;
        this.io = io;
        this.mediator = mediator;
        if (!io) {
            return;
        }

        io.on("connection", (socket) => {
            socket.on("SEND_MESSAGE", (data) => this.sendMessage(data, socket));
        });
    }

    async sendMessage({ token, message }, socket) {
        if (token && message) {

            const user = await this.db.getUserByToken(token);
            if (user) {
                this.db.sendMessage(user.id, message);
                this.io.emit("GET_MESSAGES", await this.getMessage());
                return;
            }
            this.io.to(socket.id).emit("SEND_MESSAGE", this.answer.bad(455));
            return;
        }
        this.io.to(socket.id).emit("SEND_MESSAGE", this.answer.bad(1001));
    }

    async getMessage() {
        return this.answer.good({
            messages: await this.db.getMessages(),
        });
    }
}
module.exports = Chat;
