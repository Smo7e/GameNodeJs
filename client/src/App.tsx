import React, { useState } from "react";
import { HOST, MEDIATOR } from "./config";
import { Server, Mediator, Store, TError } from "./modules";
import SignUp from "./component/SignUp/SignUp";
import Game from "./modules/Game/Game";
import Menu from "./component/Menu/Menu";
import Login from "./component/Login/Login";
import Heroes from "./component/Heroes/Heroes";
import Interface from "./component/Interface/Interface";
import Lobby from "./component/Lobby/Lobby";
import ErrorMessage from "./modules/ErrorMessage/ErrorMessage";
import Parameters from "./component/Parameters/Parameters";
import Music from "./component/Parameters/Music";
import "./App.css";

export const ServerContext = React.createContext<Server>(null!);
export const MediatorContext = React.createContext<Mediator>(null!);
export const StoreContext = React.createContext<Store>(null!);

export enum EPAGES {
    SIGNUP,
    LOGIN,
    GAME,
    MENU,
    HEROES,
    LOBBY,
    PARAMETERS,
}

const MainApp = () => {
    const [epages, setEpages] = useState<EPAGES>(EPAGES.LOGIN);
    return (
        <>
            {epages === EPAGES.LOGIN ? (
                <Login epages={setEpages} />
            ) : epages === EPAGES.SIGNUP ? (
                <SignUp epages={setEpages} />
            ) : epages === EPAGES.MENU ? (
                <Menu epages={setEpages} />
            ) : epages === EPAGES.GAME ? (
                <>
                    <Game />
                    <Interface />
                </>
            ) : epages === EPAGES.HEROES ? (
                <Heroes epages={setEpages} />
            ) : epages === EPAGES.LOBBY ? (
                <Lobby epages={setEpages} />
            ) : epages === EPAGES.PARAMETERS ? (
                <Parameters epages={setEpages} />
            ) : (
                <></>
            )}
            <ErrorMessage />
            <Music />
        </>
    );
};

const App: React.FC = () => {
    const store = new Store();
    const mediator = new Mediator(MEDIATOR);
    const server = new Server(HOST, mediator, store);

    const { SERVER_ERROR } = mediator.getEventTypes();
    mediator.subscribe(SERVER_ERROR, (data: TError) => {
        console.log(data);
    });

    return (
        <StoreContext.Provider value={store}>
            <MediatorContext.Provider value={mediator}>
                <ServerContext.Provider value={server}>
                    <MainApp />
                </ServerContext.Provider>
            </MediatorContext.Provider>
        </StoreContext.Provider>
    );
};

export default App;
