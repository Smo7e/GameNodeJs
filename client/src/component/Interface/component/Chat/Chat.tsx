import { useContext, useRef, useState } from "react";
import { useEffect } from "react";
import { MediatorContext, ServerContext } from "../../../../App";
import { TMessage } from "../../../../modules/Server/types";
import Message from "../Message/Message";
import "./Chat.css";
const Chat: React.FC = () => {
    const mediator = useContext(MediatorContext);
    const server = useContext(ServerContext);
    const [messages, setMessages] = useState<TMessage[]>([{ message: "путо", name: "тут" }]);
    const [isActiveChat, setIsActiveChat] = useState(false);
    const messageRef = useRef<HTMLInputElement>(null);

    const { GET_MESSAGES } = mediator.getEventTypes();

    const chatHandler = () => {
        setIsActiveChat(!isActiveChat);
    };
    const sendMessagesHandler = () => {
        let message = messageRef.current!.value;
        server.sendMessage(message);
    };
    useEffect(() => {
        const getMessagesHandler = (messages_: Array<TMessage>) => {
            if (messages_.length === 0) return;
            setMessages(messages_);
        };

        mediator.subscribe(GET_MESSAGES, getMessagesHandler);

        return () => {
            mediator.unsubscribe(GET_MESSAGES, getMessagesHandler);
        };
    });
    return (
        <div className="Chat-container">
            <div onClick={chatHandler} className="chat-interface">
                Чат
            </div>
            {isActiveChat ? (
                <div className="active-chat">
                    <div className="messages-list">
                        {messages.map((message_, index) =>
                            messages.length != 0 ? (
                                <Message key={index} message={message_.message} name={message_.name} />
                            ) : (
                                <></>
                            )
                        )}
                    </div>
                    <div className="send-message-container">
                        <input ref={messageRef} type="text" className="input-Message" placeholder="Ваше сообщение" />
                        <div className="Chat-image-container" onClick={sendMessagesHandler}>
                            <div className="Chat-image"></div>
                        </div>
                    </div>
                </div>
            ) : (
                <></>
            )}
        </div>
    );
};
export default Chat;
