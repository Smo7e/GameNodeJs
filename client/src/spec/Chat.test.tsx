import { render, screen } from "@testing-library/react";
import Chat from "../component/Interface/component/Chat/Chat";
import { MediatorContext, ServerContext } from "../App";
import { Mediator, Server, Store } from "../modules";
import { HOST, MEDIATOR } from "../config";

const store = new Store();
const mediator = new Mediator(MEDIATOR);
const server = new Server(HOST, mediator, store);
test("Chat snapshot", () => {
    const { asFragment } = render(
        <MediatorContext.Provider value={mediator}>
            <ServerContext.Provider value={server}>
                <Chat />
            </ServerContext.Provider>
        </MediatorContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
});
