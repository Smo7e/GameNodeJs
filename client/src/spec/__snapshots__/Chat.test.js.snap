// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat snapshot 1`] = `
<DocumentFragment>
  <div
    class="Chat-container"
  >
    <div
      class="chat-interface"
    >
      Чат
    </div>
  </div>
</DocumentFragment>
`;
